mantidproject/mantid
# Allow for Time slicing of added MDEvent data in SANS

Fixes #12698

**Context** 

We want to add event workspaces in a way that they overlap time-wise. We shift the workspaces in time such that the first time-stamp entry in the proton_charge log is identical. In addition, the user can specify a time shift to add to this initial proton_charge time-stamp overlap. Graphically this means

```
                  ---------------------------------------> time
Log1:             || |  ||  |  |||||
Log2:                                     | | ||| |||
Added:            || |  ||  |  |||||
               +  | | ||| |||
               =  |||||||||||  |||||
```

This addition and time shift occurs for the logs and the events themselves.
### For Testing:

The output was checked by Sarah Rogers, hence we have a scientifc validation for this new feature. We restrict ourselves to functional testing.
-  The required data and files can be found here: \olympic\Babylon5\Scratch\Anton\Testing\12658_Allow_for_time_slicing_of_added_event_data\Sandbox
-  Please make sure that the path to these files is set.
-  The instrument is SANS2DTUBES
-  The user file is USER_SANS2D_143ZC_2p4_4m_M4_Knowles_12mm.txt
-  The installers from Jenkins (Windows + Ubuntu) can be found here: \olympic\Babylon5\Scratch\Anton\Testing\12658_Allow_for_time_slicing_of_added_event_data
-  The data files are:
  - SANS2D00028793.nxs ==> 28793
  - SANS2D00028823.nxs ==> 28823

**Test that two workspaces are shifted on top of each other**
1. Open the ISIS SANS GUI and go to the Add Runs tab
2. Add 28793, then add 28723
3. From the dropdown select "Save As Event Data"
4. Check "OverlayEventWorkspaces" and do not specify anything for "Additional Time Shifts". Press "Sum"
   - Confirm that there are no errrors (except the warning "Error parsing duration in algorithm history entry" )
   - Confirm that  you have a file "SANS2D00028793-add.nxs" in your output directory
5. On the "Run Numbers" tab insert the path to this new file into the Scattering Sample input. Press Load
   - Confirm that two new workspaces were created
6. Open the SampleLogs and display the proton_charge log for both
   - Confirm that plotted log is continuous and that there is no gap in time. 
7. Press 1D reduction
   - Confirm that no error appears
8. Click Time Slicing on the "Run Numbers" tab. You should have only one EventWorkspace in the ADS, hence you can only slice this one.
9. Set the slices to 1 and 12  and press "Apply"
   - Confirm that there are no errors

Finally clean up. Delete the added file. Clear the workspaces(CTRL + SHIFT+L). Close the ISIS SANS GUI.

**Test that two workspaces are added without overlay**
1. Open the ISIS SANS GUI and go to the Add Runs tab
2. Add 28793, then add 28723
3. From the dropdown select "Save As Event Data"
4. Do not check "OverlayEventWorkspaces" and press "Sum"
   - Confirm that there are no errrors (except the warning "Error parsing duration in algorithm history entry" )
   - Confirm that  you have a file "SANS2D00028793-add.nxs" in your output directory
5. On the "Run Numbers" tab insert the path to this new file into the Scattering Sample input. Press Load
   - Confirm that two new workspaces were created
6. Open the SampleLogs and display the proton_charge log for both
   - Confirm that plotted log is not continuous and that there a gap in time. 
7. Press 1D reduction
   - Confirm that no error appears
8. Click Time Slicing on the "Run Numbers" tab.
9. Set the slices to 1 and 12  and press "Apply"
   - Confirm that there are no errors

Finally clean up. Delete the added file. Clear the workspace(CTRL + SHIFT+L). Close the ISIS SANS GUI.

**Test that adding workspaces with additional time shift**
1. Open the ISIS SANS GUI and go to the Add Runs tab
2. Add 28793, then add 28723
3. From the dropdown select "Save As Event Data"
4. Check "OverlayEventWorkspaces" and set 2000 in the  "Additional Time Shifts". Press "Sum"
   - Confirm that there are no errrors (except the warning "Error parsing duration in algorithm history entry" )
   - Confirm that  you have a file "SANS2D00028793-add.nxs" in your output directory
5. On the "Run Numbers" tab insert the path to this new file into the Scattering Sample input. Press Load
   - Confirm that two new workspaces were created
6. Open the SampleLogs and display the proton_charge log for both
   - Confirm that plotted log shows a small gap, should be about 200 ( this is the length (~1800) of the run minus the shift of 2000)
7. Press 1D reduction
   - Confirm that no error appears
8. Click Time Slicing on the "Run Numbers" tab. You should have only one EventWorkspace in the ADS, hence you can only slice this one.
9. Set the slices to 1 and 12  and press "Apply"
   - Confirm that a new workspace is being created with the ending _T1_T12
   - Confirm that there are no errors

Finally clean up. Delete the added file. Clear the workspace(CTRL + SHIFT+L). Close the ISIS SANS GUI.

**Test for incorrect input handling**
1. Open the ISIS SANS GUI and go to the Add Runs tab
2. Add 28793, then add 28723
3. From the dropdown select "Save As Event Data"
4. Check "OverlayEventWorkspaces" and set 2000, 300 in the  "Additional Time Shifts". Press "Sum"
   - Confirm that nothing happens and a warning is generated.
5. Set "3sdfsdfr6" as the "Additional Time Shifts". Press "Sum"
   - Confirm that nothing happens and a warning message is generated.
